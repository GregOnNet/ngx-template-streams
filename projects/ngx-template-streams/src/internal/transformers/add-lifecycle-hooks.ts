import { some } from 'fp-ts/lib/Option';
import * as ts from 'typescript';
import { findNodes, getComponentClass } from '../../utils/ast-helpers';
import { LifecycleHook, ObservableDecorator } from '../../utils/models';
import { getEventStreamDecoratorQuery, getLifecycleQuery } from './queries';

type DecoratorHooks = { [key in LifecycleHook]: Array<ObservableDecorator> };

const decoratorHooks: DecoratorHooks = {
  [LifecycleHook.ngOnDestroy]: ['ObservableEvent', 'ObservableChild', 'ObservableChildren'],
  [LifecycleHook.ngAfterViewInit]: ['ObservableChildren']
};

export function addLifecycleHooksTransformer(context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile) => {
    return some(sourceFile)
      .chain(addLifecycleHook(LifecycleHook.ngOnDestroy, context))
      .chain(addLifecycleHook(LifecycleHook.ngAfterViewInit, context))
      .getOrElse(sourceFile);
  };
}

function addLifecycleHook(lifecycle: LifecycleHook, context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile) => {
    const result = hasLifecycleHook(lifecycle, sourceFile)
      .filter(lifecycleExists => !lifecycleExists && requiresLifecycleHook(lifecycle, sourceFile))
      .chain(_ => getComponentClass(sourceFile))
      .map(updateComponentClass(lifecycle, sourceFile, context));

    return result.isNone() ? some(sourceFile) : result;
  };
}

function updateComponentClass(lifecycle: LifecycleHook, sourceFile: ts.SourceFile, context: ts.TransformationContext) {
  return (nodes: Array<ts.Node>) => {
    const visitor: ts.Visitor = (node: ts.Node) => {
      if (nodes.includes(node)) {
        const classDeclaration = node as ts.ClassDeclaration;

        const lifecycleMethod = ts.createMethod(
          [],
          undefined,
          undefined,
          ts.createIdentifier(lifecycle),
          undefined,
          undefined,
          undefined,
          undefined,
          ts.createBlock([])
        );

        return ts.updateClassDeclaration(
          classDeclaration,
          classDeclaration.decorators,
          classDeclaration.modifiers,
          classDeclaration.name,
          classDeclaration.typeParameters,
          classDeclaration.heritageClauses,
          ts.createNodeArray([...classDeclaration.members, lifecycleMethod])
        );
      }

      return ts.visitEachChild(node, visitor, context);
    };

    return ts.visitNode(sourceFile, visitor);
  };
}

function requiresLifecycleHook(lifecycle: LifecycleHook, sourceFile: ts.SourceFile) {
  const nodes = decoratorHooks[lifecycle].map(decorator =>
    findNodes<ts.ClassDeclaration>(sourceFile, getEventStreamDecoratorQuery(decorator))
  );

  return nodes.some(declarations => declarations.isSome());
}

function hasLifecycleHook(lifecycle: LifecycleHook, sourceFile: ts.SourceFile) {
  const nodes = findNodes<ts.ClassDeclaration>(sourceFile, getLifecycleQuery(lifecycle));
  return some(nodes.isSome());
}
