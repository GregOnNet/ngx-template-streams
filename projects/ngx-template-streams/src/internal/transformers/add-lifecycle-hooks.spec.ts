import { createFileContent } from '../../testing/test-helpers';
import { transformContent } from '../../testing/ts-compiler';
import * as transformer from './add-lifecycle-hooks';

describe('addLifecycleHooksTransformer', () => {
  let transformerSpy: jest.SpyInstance;

  beforeEach(() => {
    transformerSpy = jest.spyOn(transformer, 'addLifecycleHooksTransformer');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should not apply transformation if the file is empty', () => {
    const content = '';

    const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

    expect(output).toMatchSnapshot();
    expect(transformerSpy).toHaveBeenCalled();
  });

  it('should not apply transformation if class doesn not contain known decorators', () => {
    const content = createFileContent`
      @Component()
      class TestComponent {}
    `;

    const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

    expect(output).toMatchSnapshot();
    expect(transformerSpy).toHaveBeenCalled();
  });

  describe('ngOnDestroy', () => {
    it('should not add ngOnDestroy for unknown decorators', () => {
      const content = createFileContent`
          @Component()
          class TestComponent {
            @Input() myInput: number;
            @Output() myOutput = new EventEmitter();
            @ViewChild('test', { static: true }) testElement;
          }
        `;

      const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });

    test.each(['ObservableEvent', 'ObservableChild', 'ObservableChildren'])(
      'should add ngOnDestroy if it does not exist › %s',
      decorator => {
        const content = createFileContent`
          @Component()
          class TestComponent {
            @${decorator}()
            stream$: Observable<any>;
          }
        `;

        const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

        expect(output).toMatchSnapshot();
        expect(transformerSpy).toHaveBeenCalled();
      }
    );

    test.each(['ObservableEvent', 'ObservableChild', 'ObservableChildren'])(
      'should not add ngOnDestroy if it exists › %s',
      decorator => {
        const content = createFileContent`
          @Component()
          class TestComponent {
            @${decorator}()
            stream$: Observable<any>;

            ngOnDestroy() {
              console.log('test');
            }
          }
        `;

        const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

        expect(output).toMatchSnapshot();
        expect(transformerSpy).toHaveBeenCalled();
      }
    );
  });

  describe('ngAfterViewInit', () => {
    it('should add ngAfterViewInit if it does not exist', () => {
      const content = createFileContent`
        @Component()
        class TestComponent {
          @ObservableChildren()
          stream$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });

    it('should not add ngAfterViewInit if it exists', () => {
      const content = createFileContent`
        @Component()
        class TestComponent {
          @ObservableChildren()
          stream$: Observable<any>;

          ngAfterViewInit() {
            console.log('test');
          }
        }
      `;

      const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });

    test.each(['ObservableEvent', 'ObservableChild'])('should not add ngAfterViewInit › %s', decorator => {
      const content = createFileContent`
        @Component()
        class TestComponent {
          @${decorator}()
          stream$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addLifecycleHooksTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });
  });
});
