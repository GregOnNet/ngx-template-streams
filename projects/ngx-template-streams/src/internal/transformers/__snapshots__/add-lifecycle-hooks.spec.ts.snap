// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addLifecycleHooksTransformer ngAfterViewInit should add ngAfterViewInit if it does not exist 1`] = `
"@Component()
class TestComponent {
    @ObservableChildren()
    stream$: Observable<any>;
    ngOnDestroy() { }
    ngAfterViewInit() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngAfterViewInit should not add ngAfterViewInit if it exists 1`] = `
"@Component()
class TestComponent {
    @ObservableChildren()
    stream$: Observable<any>;
    ngAfterViewInit() {
        console.log('test');
    }
    ngOnDestroy() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngAfterViewInit should not add ngAfterViewInit › ObservableChild 1`] = `
"@Component()
class TestComponent {
    @ObservableChild()
    stream$: Observable<any>;
    ngOnDestroy() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngAfterViewInit should not add ngAfterViewInit › ObservableEvent 1`] = `
"@Component()
class TestComponent {
    @ObservableEvent()
    stream$: Observable<any>;
    ngOnDestroy() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngOnDestroy should add ngOnDestroy if it does not exist › ObservableChild 1`] = `
"@Component()
class TestComponent {
    @ObservableChild()
    stream$: Observable<any>;
    ngOnDestroy() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngOnDestroy should add ngOnDestroy if it does not exist › ObservableChildren 1`] = `
"@Component()
class TestComponent {
    @ObservableChildren()
    stream$: Observable<any>;
    ngOnDestroy() { }
    ngAfterViewInit() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngOnDestroy should add ngOnDestroy if it does not exist › ObservableEvent 1`] = `
"@Component()
class TestComponent {
    @ObservableEvent()
    stream$: Observable<any>;
    ngOnDestroy() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngOnDestroy should not add ngOnDestroy for unknown decorators 1`] = `
"@Component()
class TestComponent {
    @Input()
    myInput: number;
    @Output()
    myOutput = new EventEmitter();
    @ViewChild('test', { static: true })
    testElement;
}
"
`;

exports[`addLifecycleHooksTransformer ngOnDestroy should not add ngOnDestroy if it exists › ObservableChild 1`] = `
"@Component()
class TestComponent {
    @ObservableChild()
    stream$: Observable<any>;
    ngOnDestroy() {
        console.log('test');
    }
}
"
`;

exports[`addLifecycleHooksTransformer ngOnDestroy should not add ngOnDestroy if it exists › ObservableChildren 1`] = `
"@Component()
class TestComponent {
    @ObservableChildren()
    stream$: Observable<any>;
    ngOnDestroy() {
        console.log('test');
    }
    ngAfterViewInit() { }
}
"
`;

exports[`addLifecycleHooksTransformer ngOnDestroy should not add ngOnDestroy if it exists › ObservableEvent 1`] = `
"@Component()
class TestComponent {
    @ObservableEvent()
    stream$: Observable<any>;
    ngOnDestroy() {
        console.log('test');
    }
}
"
`;

exports[`addLifecycleHooksTransformer should not apply transformation if class doesn not contain known decorators 1`] = `
"@Component()
class TestComponent {
}
"
`;

exports[`addLifecycleHooksTransformer should not apply transformation if the file is empty 1`] = `""`;
